codalab/rest/bundle.py

```diff
@@ -77,3 +86,81 @@ def get_blob(uuid, path=''):
     response.set_header('Content-Encoding', encoding or 'identity')
 
     return fileobj
+
+
+@put('/bundle/<uuid:re:%s>/contents/blob/' % spec_util.UUID_STR,
+     apply=AuthenticatedPlugin())
+def put_blob(uuid):
+    try:
+        check_bundles_have_all_permission(local.model, request.user, [uuid])
+    except PermissionError as e:
+        abort(httplib.FORBIDDEN, e.message)
+
+    bundle = safe_get_bundle(uuid)
+
+    # If this bundle already has data, remove it.
+    bundle_path = os.path.realpath(local.bundle_store.get_bundle_location(uuid))
+    if os.path.exists(bundle_path):
+        local.bundle_store.cleanup(uuid, dry_run=False)
+        bundle_update = {
+            'data_hash': None,
+        }
+        local.model.update_bundle(bundle, bundle_update)
+
+    # Figure out what kind of upload it is.
+    tar_archive = False
+    single_file_archive = False
+    if 'Content-Disposition' in request.headers:
```

Comment:
I have made the function work like the existing upload mechanism. See what happens when you try to upload a tar archive. I don't see the point of adding a query parameter until someone starts using it. Otherwise, it's just dead code. When Andrew does the migration to this new API, he's not going to use the query parameter. These is simply no place to specify for the user to specify it.

I've documented how the Content-Disposition header should be populated. For reference, when you currently try to upload a file, here's what gets sent in the request (not including the actual binary data). Note the filename in the first section. That is the filename on my computer.

------WebKitFormBoundaryBDtYlm2Ql0Cde31F
Content-Disposition: form-data; name="file"; filename="bundle.tar.gz"
Content-Type: application/gzip


------WebKitFormBoundaryBDtYlm2Ql0Cde31F
Content-Disposition: form-data; name="bundle_type"

dataset
------WebKitFormBoundaryBDtYlm2Ql0Cde31F
Content-Disposition: form-data; name="worksheet_uuid"

0x13894eb2358b437eb2df8bd371599fd1
------WebKitFormBoundaryBDtYlm2Ql0Cde31F--